File: examples/python/error-catalog.py
"""MissionInbox Python SDK — error catalog.

Deliberately triggers each error the SDK maps, prints the raw HTTP response
body observed, and verifies the SDK's exception classification. Companion to
``main.py``; run separately via ``python error-catalog.py``.
"""

from __future__ import annotations

import os
import sys
import time
from typing import Any, Callable

from dotenv import load_dotenv

from missioninbox import (
    AuthenticationError,
    ConflictError,
    MissionInbox,
    MissionInboxError,
    NetworkError,
    NotFoundError,
    UnregisteredSenderError,
    UnverifiedDomainError,
    ValidationError,
)

load_dotenv()

api_key = os.environ.get("MI_API_KEY", "")
base_url = os.environ.get("MI_API_URL", "")
test_sender = os.environ.get("MI_TEST_SENDER", "")
test_to = os.environ.get("MI_TEST_TO", "")
test_domain = os.environ.get("MI_TEST_DOMAIN", "")

if not api_key or not base_url:
    print("MI_API_KEY and MI_API_URL are required.", file=sys.stderr)
    sys.exit(2)

mi = MissionInbox(api_key=api_key, base_url=base_url, max_retries=0)
ts = int(time.time() * 1000)
can_trigger = bool(test_sender and test_to and test_domain)

results: list[dict[str, Any]] = []


def trigger(id_: int, name: str, expected_cls: type[BaseException], fn: Callable[[], Any]) -> None:
    expected = expected_cls.__name__
    print(f"\n━━━ {id_}. {expected}: {name} ━━━")
    try:
        r = fn()
        print(f"  ✗ Expected {expected}, got success: {str(r)[:120]}")
        results.append({"id": id_, "name": name, "expected": expected, "actual": "no-throw", "status": 0, "pass": False})
    except Exception as err:
        actual = type(err).__name__
        status = err.status if isinstance(err, MissionInboxError) else 0
        body = err.body if isinstance(err, MissionInboxError) else None
        passed = isinstance(err, expected_cls)
        mark = "✓" if passed else "✗"
        print(f"  {mark} Actual class:  {actual}" + ("" if passed else f"  (expected {expected})"))
        print(f"     HTTP status:  {status}")
        print(f"     Response body: {body!r}")
        print(f"     Message:       {str(err)[:160]}")
        results.append({"id": id_, "name": name, "expected": expected, "actual": actual, "status": status, "pass": passed})


print(f"MissionInbox error-catalog run — base={base_url}")
if not can_trigger:
    print("  (MI_TEST_SENDER / TO / DOMAIN not set — will skip triggers that need them)")

# 1. AuthenticationError
bad_key = MissionInbox(api_key="obviously-wrong-key", base_url=base_url, max_retries=0)
trigger(1, "wrong API key", AuthenticationError, lambda: bad_key.emails.get_send_limit())

# 2. ValidationError — no recipient
if can_trigger:
    trigger(2, "send with no recipient", ValidationError, lambda: mi.emails.send(
        from_=test_sender,
        subject="no recipient",
        text="should fail validation",
    ))
else:
    print("\n━━━ 2. ValidationError: send with no recipient ━━━\n  (skipped)")

# 3. ValidationError — no body
if can_trigger:
    trigger(3, "send with no body", ValidationError, lambda: mi.emails.send(
        from_=test_sender,
        to=test_to,
        subject="no body",
    ))
else:
    print("\n━━━ 3. ValidationError: send with no body ━━━\n  (skipped)")

# 4. UnregisteredSenderError
if can_trigger:
    trigger(4, "send from unregistered address", UnregisteredSenderError, lambda: mi.emails.send(
        from_=f"never-registered-{ts}@example.invalid",
        to=test_to,
        subject="unregistered",
        text="should fail",
    ))
else:
    print("\n━━━ 4. UnregisteredSenderError ━━━\n  (skipped)")

# 5. NotFoundError
trigger(5, "fetch non-existent sending identifier", NotFoundError,
        lambda: mi.sending_identifiers.get("00000000-0000-0000-0000-000000000000"))

# 6. ConflictError — register identifier twice
if can_trigger:
    test_addr = f"sdk-error-catalog-{ts}@{test_sender.split('@', 1)[1]}"
    first_id = None
    try:
        created = mi.sending_identifiers.create(
            email_address=test_addr,
            display_name="SDK error catalog — safe to delete",
        )
        first_id = created["id"]
        trigger(6, "register identifier that already exists", ConflictError,
                lambda: mi.sending_identifiers.create(email_address=test_addr, display_name="duplicate"))
    except Exception as err:
        print("\n━━━ 6. ConflictError: register identifier twice ━━━")
        print(f"  ✗ Setup failed: {err}")
    finally:
        if first_id:
            try:
                mi.sending_identifiers.delete(first_id)
            except Exception:
                pass
else:
    print("\n━━━ 6. ConflictError ━━━\n  (skipped)")

# 7. UnverifiedDomainError
if can_trigger:
    sub_domain = f"sdk-error-{ts}.{test_domain}"
    test_from = f"sender@{sub_domain}"
    identifier_id = None
    try:
        created = mi.domains.bulk_create(domains=[{"domainName": sub_domain}])
        mi.tasks.wait_for(created["taskId"], poll_interval=3.0, timeout=30.0)
        identifier = mi.sending_identifiers.create(email_address=test_from)
        identifier_id = identifier["id"]
        trigger(7, "send from unverified-domain identifier", UnverifiedDomainError,
                lambda: mi.emails.send(from_=test_from, to=test_to, subject="unverified", text="should fail"))
    except Exception as err:
        print("\n━━━ 7. UnverifiedDomainError ━━━")
        print(f"  ✗ Setup failed: {err}")
    finally:
        if identifier_id:
            try:
                mi.sending_identifiers.delete(identifier_id)
            except Exception:
                pass
        try:
            mi.domains.bulk_delete(domain_names=[sub_domain])
        except Exception:
            pass
else:
    print("\n━━━ 7. UnverifiedDomainError ━━━\n  (skipped)")

# 8. NetworkError
unreachable = MissionInbox(api_key=api_key, base_url="https://127.0.0.1:1", max_retries=0, timeout=2.0)
trigger(8, "unreachable host", NetworkError, lambda: unreachable.health.check())

# Summary
print("\n━━━ Summary ━━━")
passes = sum(1 for r in results if r["pass"])
print(f"  {passes}/{len(results)} exception mappings correct")
for r in results:
    mark = "✓" if r["pass"] else "✗"
    print(f"  {mark} #{r['id']} [{r['status']}] {r['expected']}: {r['name']}")
print()
print("Not tested here (need special account state or would burden staging):")
print("  SubscriptionInactiveError — inactive account")
print("  SendLimitExceededError    — Free plan hitting 20/day cap")
print("  DomainBlacklistedError    — blacklisted domain")
print("  RateLimitError            — sustained request volume")
print("  ServerError               — API 5xx")
print("  SendError                 — SES rejection")
print("  PermissionError (base)    — non-specific 403")

File: examples/node/error-catalog.ts
/**
 * MissionInbox Node SDK — error catalog.
 *
 * Deliberately triggers each error the SDK maps, prints the raw HTTP
 * response body observed, and verifies the SDK's exception classification.
 * Companion to `main.ts`; run separately via `npm run errors`.
 *
 * Env vars (same as main.ts):
 *   MI_API_KEY, MI_API_URL — required
 *   MI_TEST_SENDER, MI_TEST_TO, MI_TEST_DOMAIN — required for triggers 3–7
 */

import 'dotenv/config';
import {
  AuthenticationError,
  ConflictError,
  MissionInbox,
  MissionInboxError,
  NetworkError,
  NotFoundError,
  UnregisteredSenderError,
  UnverifiedDomainError,
  ValidationError,
} from '@missioninbox/sdk';

const {
  MI_API_KEY,
  MI_API_URL,
  MI_TEST_SENDER,
  MI_TEST_TO,
  MI_TEST_DOMAIN,
} = process.env;

if (!MI_API_KEY || !MI_API_URL) {
  console.error('MI_API_KEY and MI_API_URL are required.');
  process.exit(2);
}

const mi = new MissionInbox({
  apiKey: MI_API_KEY,
  baseUrl: MI_API_URL,
  maxRetries: 0,
});

const ts = Date.now();
const results: Array<{
  id: number;
  name: string;
  expected: string;
  actual: string;
  status: number;
  body: unknown;
  pass: boolean;
}> = [];

async function trigger(
  id: number,
  name: string,
  expectedClass: new (...args: never[]) => Error,
  fn: () => Promise<unknown>,
): Promise<void> {
  const expected = expectedClass.name;
  console.log(`\n━━━ ${id}. ${expected}: ${name} ━━━`);
  try {
    const r = await fn();
    console.log(`  ✗ Expected ${expected}, got success:`, JSON.stringify(r).slice(0, 120));
    results.push({ id, name, expected, actual: 'no-throw', status: 0, body: r, pass: false });
  } catch (err) {
    const actual = err instanceof Error ? err.constructor.name : typeof err;
    const status = err instanceof MissionInboxError ? err.status : 0;
    const body = err instanceof MissionInboxError ? err.body : undefined;
    const pass = err instanceof expectedClass;
    console.log(`  ${pass ? '✓' : '✗'} Actual class:  ${actual}${pass ? '' : `  (expected ${expected})`}`);
    console.log(`     HTTP status:  ${status}`);
    console.log(`     Response body: ${JSON.stringify(body)}`);
    if (err instanceof Error && err.message) {
      console.log(`     Message:       ${err.message.slice(0, 160)}`);
    }
    results.push({ id, name, expected, actual, status, body, pass });
  }
}

async function main(): Promise<void> {
  console.log(`MissionInbox error-catalog run — base=${MI_API_URL}`);
  const canTrigger = Boolean(MI_TEST_SENDER && MI_TEST_TO && MI_TEST_DOMAIN);
  if (!canTrigger) {
    console.log('  (MI_TEST_SENDER / TO / DOMAIN not set — will skip triggers that need them)');
  }

  // 1. AuthenticationError (401) — wrong API key
  const badKey = new MissionInbox({
    apiKey: 'obviously-wrong-key',
    baseUrl: MI_API_URL!,
    maxRetries: 0,
  });
  await trigger(1, 'wrong API key', AuthenticationError, () => badKey.emails.getSendLimit());

  // 2a. ValidationError (400) — missing recipient
  if (canTrigger) {
    await trigger(2, 'send with no recipient', ValidationError, () =>
      mi.emails.send({
        from: MI_TEST_SENDER!,
        subject: 'no recipient',
        text: 'should fail validation',
      } as never),
    );
  } else {
    console.log('\n━━━ 2. ValidationError: send with no recipient ━━━\n  (skipped — set MI_TEST_SENDER)');
  }

  // 2b. ValidationError (400) — missing body
  if (canTrigger) {
    await trigger(3, 'send with no body', ValidationError, () =>
      mi.emails.send({
        from: MI_TEST_SENDER!,
        to: MI_TEST_TO!,
        subject: 'no body content',
      } as never),
    );
  } else {
    console.log('\n━━━ 3. ValidationError: send with no body ━━━\n  (skipped)');
  }

  // 3. UnregisteredSenderError (403)
  if (canTrigger) {
    await trigger(4, 'send from unregistered address', UnregisteredSenderError, () =>
      mi.emails.send({
        from: `never-registered-${ts}@example.invalid`,
        to: MI_TEST_TO!,
        subject: 'unregistered from',
        text: 'should fail',
      }),
    );
  } else {
    console.log('\n━━━ 4. UnregisteredSenderError: send from unregistered address ━━━\n  (skipped)');
  }

  // 4. NotFoundError (404) — fetch non-existent sending identifier
  await trigger(5, 'fetch non-existent sending identifier', NotFoundError, () =>
    mi.sendingIdentifiers.get('00000000-0000-0000-0000-000000000000'),
  );

  // 5. ConflictError (409) — register the same identifier twice
  if (canTrigger) {
    const testAddr = `sdk-error-catalog-${ts}@${MI_TEST_SENDER!.split('@')[1]}`;
    let firstId: string | undefined;
    try {
      const created = await mi.sendingIdentifiers.create({
        emailAddress: testAddr,
        displayName: 'SDK error catalog — safe to delete',
      });
      firstId = created.id;
      // Now trigger the conflict:
      await trigger(6, 'register identifier that already exists', ConflictError, () =>
        mi.sendingIdentifiers.create({
          emailAddress: testAddr,
          displayName: 'duplicate',
        }),
      );
    } catch (err) {
      console.log('\n━━━ 6. ConflictError: register identifier twice ━━━');
      console.log(`  ✗ Setup failed:`, err instanceof Error ? err.message : String(err));
    } finally {
      if (firstId) {
        try {
          await mi.sendingIdentifiers.delete(firstId);
        } catch {
          /* best effort cleanup */
        }
      }
    }
  } else {
    console.log('\n━━━ 6. ConflictError: register identifier twice ━━━\n  (skipped)');
  }

  // 6. UnverifiedDomainError (403) — create sub-domain, register identifier, try to send
  if (canTrigger && MI_TEST_DOMAIN) {
    const subDomain = `sdk-error-${ts}.${MI_TEST_DOMAIN}`;
    const testFrom = `sender@${subDomain}`;
    let taskId: string | undefined;
    let identifierId: string | undefined;
    try {
      const created = await mi.domains.bulkCreate({ domains: [{ domainName: subDomain }] });
      taskId = created.taskId;
      await mi.tasks.waitFor(taskId, { pollInterval: 3_000, timeout: 30_000 });
      const identifier = await mi.sendingIdentifiers.create({ emailAddress: testFrom });
      identifierId = identifier.id;
      await trigger(7, 'send from unverified-domain identifier', UnverifiedDomainError, () =>
        mi.emails.send({
          from: testFrom,
          to: MI_TEST_TO!,
          subject: 'unverified domain',
          text: 'should fail',
        }),
      );
    } catch (err) {
      console.log('\n━━━ 7. UnverifiedDomainError: send from unverified-domain identifier ━━━');
      console.log(`  ✗ Setup failed:`, err instanceof Error ? err.message : String(err));
    } finally {
      if (identifierId) {
        try {
          await mi.sendingIdentifiers.delete(identifierId);
        } catch {
          /* ignore */
        }
      }
      try {
        await mi.domains.bulkDelete({ domainNames: [subDomain] });
      } catch {
        /* ignore */
      }
    }
  } else {
    console.log('\n━━━ 7. UnverifiedDomainError: unverified domain ━━━\n  (skipped)');
  }

  // 7. NetworkError — client-side, unroutable host
  const unreachable = new MissionInbox({
    apiKey: MI_API_KEY!,
    baseUrl: 'https://127.0.0.1:1',
    maxRetries: 0,
    timeout: 2_000,
  });
  await trigger(8, 'unreachable host', NetworkError, () => unreachable.health.check());

  // ─── Summary ─────────────────────────────────────────────────────────────
  console.log('\n━━━ Summary ━━━');
  const passes = results.filter((r) => r.pass).length;
  console.log(`  ${passes}/${results.length} exception mappings correct`);
  for (const r of results) {
    const mark = r.pass ? '✓' : '✗';
    console.log(`  ${mark} #${r.id} [${r.status}] ${r.expected}: ${r.name}`);
  }
  console.log('');
  console.log('Not tested here (need special account state or would burden staging):');
  console.log('  SubscriptionInactiveError — inactive account');
  console.log('  SendLimitExceededError    — Free plan hitting 20/day cap');
  console.log('  DomainBlacklistedError    — blacklisted domain');
  console.log('  RateLimitError            — sustained request volume');
  console.log('  ServerError               — API 5xx');
  console.log('  SendError                 — SES rejection with a valid-looking recipient');
  console.log('  PermissionError (base)    — non-specific 403');
}

main().catch((err) => {
  console.error('\nUnhandled error:', err);
  process.exit(1);
});
